src/Common/FileSegment.cpp

```diff
@@ -684,29 +730,30 @@ bool FileSegment::hasFinalizedState() const
         || download_state == State::SKIP_CACHE;
 }
 
-void FileSegment::detach(std::lock_guard<std::mutex> & cache_lock, std::lock_guard<std::mutex> & segment_lock)
+void FileSegment::detach(
+    std::lock_guard<std::mutex> & /* cache_lock */,
+    std::lock_guard<std::mutex> & segment_lock)
 {
-    if (detached)
+    if (is_detached)
         return;
 
     markAsDetached(segment_lock);
+    download_state = State::PARTIALLY_DOWNLOADED_NO_CONTINUATION;
+    downloader_id.clear();
```

Comment:
the most important change here, though its absence should not affect anything because now detached status can be in 2 cases, which do not do any complex logic and check exactly for detached status:
1) there is only 1 remaining file segment holder && it does not need this segment anymore && this file segment was in cache and needs to be removed
2) in `read_from_cache_if_exists_otherwise_bypass_cache` case (`get` method can create them)

So this change not required, but it is more correct to have it (so the file segment state becomes consistent). Other remaining changes in this pr are just assertions and adding logical errors in some places.